Move WEBGL_CONSTANTS and THREE_TO_WEBGL constants out from GLTFExporter and GLTFLoader

fernandojsg opened the pull request:
Not so happy with the name `THREEToWebGL.js` and `WebGLToTHREE` any suggestions? :)
I believe most of the mappings from `WebGLUtils` (https://github.com/mrdoob/three.js/blob/dev/src/renderers/webgl/WebGLUtils.js#L13-L35 ...) could be replaced by this mappings constants as I find it more handy to use than need to create a `WebGLUtils` instance with a `renderer` and `extensions`. Maybe in another PR we could just reuse the constants mapping on the WebGLUtils function and just leave the code related to extensions there?

fernandojsg committed:
Move WEBGL_CONSTANTS from GLTFExporter/Loader to WebGLConstants

fernandojsg committed:
Move THREE_TO_WEBGL out from GLTFExporter

fernandojsg committed:
Added WEBGL_TO_THREE mappings and apply it on GLTFLoader

takahirox commented:
The variables are glTF specific, correct?
I'm feeling like the variable names (of file names) should indicate they're for glTF.
So add prefix GLTF like `GLTF_WEBGL_CONSTANTS`?

fernandojsg commented:
@takahirox these are not gltf specific, these are generic WEBGL constants (https://developer.mozilla.org/en-US/docs/Web/API/WebGL_API/Constants) that can be used in other places not just in the gltf loader.

takahirox commented:
Oh, really. I didn't know that, thanks!

fernandojsg commented:
Yes, the point was to avoid sending a `renderer`'s context to the exporter (As discussed in https://github.com/mrdoob/three.js/pull/11964)

takahirox commented:
I see I see, that makes sense!

mrdoob commented on src/Three.js:
```diff
@@ -156,5 +156,8 @@ export { Curve } from './extras/core/Curve.js';
 export { ShapeUtils } from './extras/ShapeUtils.js';
 export { SceneUtils } from './extras/SceneUtils.js';
 export { WebGLUtils } from './renderers/webgl/WebGLUtils.js';
+export { WEBGL_CONSTANTS } from './renderers/webgl/WebGLConstants.js';
```
I think I would keep this as:

```js
export { WebGLConstants } from './renderers/webgl/WebGLConstants.js';
```

Also, I think I would add a `fromGL()` and `toGL()` methods that does the `THREEToWebGL` and WebGLToTHREE` conversions. Just to keep everything in one file.

fernandojsg replied:
@mrdoob Do you mean including and export `fromGL`, `toGL` and `WebGLConstants` from `WebGLConstants.js`?

fernandojsg replied:
If that's the case, maybe rename the export to `constants` so you could use it as:
```js
THREE.WebGLConstants.constants.POINTS
THREE.WebGLConstants.fromGL(gl.POINTS)
THREE.WebGLConstants.toGL(THREE.FrontSide)
```

mrdoob replied:
I like this API style better.

```js
var constants = new THREE.WebGLConstants();
constants.POINTS;
constants.fromGL(gl.POINTS);
constants.toGL(THREE.FrontSide);
```

But I'm not 100%. Maybe `fromGL` and `toGL` could go on `WebGLUtils` instead.

fernandojsg replied:
I don't really like the need to create a new instance just to query a constant. I would prefer to access it directly:
```js
THREE.WebGLConstants.POINTS;
```

The same for the `fromGL` and `toGL` that I believe too that they should be part of  `WebGLUtils`
```js
THREE.WebGLUtils.fromGL(gl.POINTS)
THREE.WebGLUtils.toGL(THREE.FrontSide)
```

fernandojsg replied:
@mrdoob I've updated `WebGLUtils` and `WebGLConstants` (still WIP) but to show a possible implementation of this last option

fernandojsg replied:
Eventually `convert` and `toGL` will become the same function. In fact right now with the current modifications `convert` doesn't need a `gl` context anymore just the `extensions` dependency.

fernandojsg committed:
Rename WEBGL_CONSTANTS to WebGLConstants

fernandojsg committed:
Moved fromGL and toGL to WebGLUtils

fernandojsg committed:
Replaced WebGLUtils.convert with toGL

fernandojsg commented:
@mrdoob I've updated the `WebGLUtils` to export `toGL` and `fromGL`. I've replaced `convert` by `toGL` in the rest of the code as it was just duplicated code and the need to create a new instance and the dependencies from a gl context and extensions are not needed anymore as we're just exposing constants that we've in the `WebGLConstants` file.
I didn't change yet the `GLTFLoader/Exporter` until you'd validate the code.

Taking a look at `constants.js`, I believe I could export the `WebGLConstants` in a similar way instead of just exporting the whole object if you prefer so.

takahirox commented:
Will this PR be merged soon?
I'm making PRs which use constants.
I'll wait for this PR being merged if it'll be soon.

fernandojsg committed:
GLTFExporter: Rebased and fixed WebGLUtils use on the exporter

fernandojsg commented:
Rebased!
I export `WebGLUtils` object that contains the `toGL` and `fromGL` functions that are basically just an alias to access to the mapping dictionaries.
I believe it's more intuitive and easy to use this way instead of creating a new instance of the object everytime you need to use, the same way as we don't do `var math = new Math(); math.clamp(x)` every time we need to use a math function.
@mrdoob friendly bump :wave: :angel:

fernandojsg commented:
I'm not sure why the `lgtm` tests are failing they were ok I did a rebase and they seem to be broken since then ¯\_(ツ)_/¯

takahirox commented:
`lgtm` tests are fine now 😄

fernandojsg commented:
@takahirox yep, I just rebased it and pushed it again :)

takahirox commented:
When will we merge this PR?

mrdoob commented on examples/js/exporters/GLTFExporter.js:
```diff
@@ -373,10 +336,10 @@ THREE.GLTFExporter.prototype = {
-				minFilter: THREE_TO_WEBGL[ map.minFilter ],
-				wrapS: THREE_TO_WEBGL[ map.wrapS ],
-				wrapT: THREE_TO_WEBGL[ map.wrapT ]
+				magFilter: THREE.WebGLUtils.toGL[ map.magFilter ],
```
`toGL` is a function 😅

fernandojsg replied:
fixed 👼

mrdoob commented on examples/js/loaders/GLTFLoader.js:
```diff
@@ -1381,7 +1264,7 @@ THREE.GLTFLoader = ( function () {
 
 				var itemSize = WEBGL_TYPE_SIZES[ accessor.type ];
-				var TypedArray = WEBGL_COMPONENT_TYPES[ accessor.componentType ];
+				var TypedArray = THREE.WEBGL_TO_THREE[ accessor.componentType ];
```
Wouldn't it be better to use the same approach everywhere? `THREE.WebGLUtils.fromGL()`...

fernandojsg replied:
fixed

mrdoob commented on src/renderers/WebGLRenderer.js:
```diff
@@ -31,7 +31,7 @@ import { Vector3 } from '../math/Vector3.js';
 import { Frustum } from '../math/Frustum.js';
 import { Vector4 } from '../math/Vector4.js';
-import { WebGLUtils } from './webgl/WebGLUtils.js';
+import { WebGLUtils } from './webgl/WebGLUtils';
```
Lets keep `WebGLUtils.js`.

fernandojsg replied:
fixed

mrdoob commented on src/renderers/webgl/WebGLUtils.js:
```diff
@@ -1,140 +1,212 @@
+//------------------------------------------------------------------------------
+var THREE_TO_WEBGL = {
+
+	[ Constants.NearestFilter ]: WebGLConstants.NEAREST,
```
Was not aware of this `[ variable ]` thing. Is it well supported?

fernandojsg replied:
It's ES6 computed literals. I thought gulp would do the transform itself but it doesn't seems so :\

fernandojsg replied:
The ugly but compatible version is to put the value of the constant as is, or try to use a gulp pluging to convert it to its literal value when creating the bundle. @mrdoob thoughts?

fernandojsg replied:
My fault, this is es2015 and it's supported in all the browsers, but on firefox we need to add extra parenthesis: `[ (value) ]`. I'll fix it

fernandojsg replied:
actually, on the latest version of nightly it works as expected :)

fernandojsg replied:
@mrdoob my fault, I was testing it incorrectly on firefox O:) it's working as expected in all the browsers, also tested on mobile (android & ios).

fernandojsg committed:
Fixes few issues and rebased

mrdoob commented:
Closing this for now.
